src/toolkits/object_detection/object_detector.cpp

```diff
@@ -990,6 +1001,42 @@ std::unique_ptr<data_iterator> object_detector::create_iterator(
       iterator_params.annotation_position = data_iterator::annotation_position_enum::BOTTOM_LEFT;
   }
 
+  // Append empty annotations if no annotations are found
+  if (!iterator_params.data.contains_column(
+          iterator_params.annotations_column_name)) {
+    std::vector<flexible_type> dummy;
+    for (size_t i = 0; i < iterator_params.data.size(); i++) {
+      dummy.push_back(flex_list());
+    }
+    gl_sarray empty_ann(dummy, flex_type_enum::LIST);
+    iterator_params.data.add_column(empty_ann,
+                                    iterator_params.annotations_column_name);
+    iterator_params.has_annotations = false;
+  }
+
+  // Check annotations type. If it is a single dictionary, put it into a list.
+  std::vector<flexible_type> list_type_annotations;
+  for (size_t i = 0; i < iterator_params.data.size(); i++) {
+    flexible_type annotation =
+        iterator_params.data[iterator_params.annotations_column_name][i];
+    if (annotation.get_type() == flex_type_enum::LIST) {
+      list_type_annotations.push_back(annotation);
+    } else if (annotation.get_type() == flex_type_enum::DICT) {
+      flex_list dummy{annotation};
+      list_type_annotations.push_back(dummy);
+    } else if (annotation.get_type() == flex_type_enum::UNDEFINED) {
+      // When the annotation is None
+      list_type_annotations.push_back(flex_list());
+    } else {
+      log_and_throw(
+          "Invalid annotation! Expect list of dictionary, a singel "
+          "dictionary!");
+    }
+  }
+  gl_sarray ann_sarray(list_type_annotations, flex_type_enum::LIST);
+  iterator_params.data.replace_add_column(
+      ann_sarray, iterator_params.annotations_column_name);
```

Comment:
Could we handle this on the Python side? I think it makes the most sense for the C++ code to be more strict about types.

Also, we shouldn't be using random access to seek into SArray or SFrame values. This is extremely inefficient! Talk to @Jarvi-Izana or @TobyRoseman about using `apply` or other parallel methods. You should also be able to query the column's type upfront, not query each individual element